src/main/java/com/puppycrawl/tools/checkstyle/site/PropertiesMacro.java

```diff
@@ -0,0 +1,403 @@
+        writeTablePropertiesRows(sink);
+        sink.rawText(INDENT_LEVEL_10);
+        sink.tableRows_();
+        sink.table_();
+        sink.setInsertNewline(true);
+    }
+
+    /**
+     * Writes the table header row with 5 columns - name, description, type, default value, since.
+     *
+     * @param sink sink to write to.
+     */
+    private static void writeTableHeaderRow(Sink sink) {
+        sink.tableRow();
+        writeTableHeaderCell(sink, "name");
+        writeTableHeaderCell(sink, "description");
+        writeTableHeaderCell(sink, "type");
+        writeTableHeaderCell(sink, "default value");
+        writeTableHeaderCell(sink, "since");
+        sink.rawText(INDENT_LEVEL_12);
+        sink.tableRow_();
+    }
+
+    /**
+     * Writes a table header cell with the given text.
+     *
+     * @param sink sink to write to.
+     * @param text the text to write.
+     */
+    private static void writeTableHeaderCell(Sink sink, String text) {
+        sink.rawText(INDENT_LEVEL_14);
+        sink.tableHeaderCell();
+        sink.text(text);
+        sink.tableHeaderCell_();
+    }
+
+    /**
+     * Writes the rows of the table with the 5 columns - name, description, type, default value,
+     * since. Each row corresponds to a property of the module.
+     *
+     * @param sink sink to write to.
+     * @throws MacroExecutionException if an error occurs during writing.
+     */
+    private static void writeTablePropertiesRows(Sink sink)
+            throws MacroExecutionException {
+        final Object instance = SiteUtil.getModuleInstance(currentModuleName);
+        final Class<?> clss = instance.getClass();
+
+        final Set<String> properties = SiteUtil.getPropertiesForDocumentation(clss, instance);
+        final Map<String, DetailNode> propertiesJavadocs = SiteUtil
+                .getPropertiesJavadocs(properties, currentModuleName, currentModuleFile);
+
+        for (String property : properties) {
+            final DetailNode propertyJavadoc = propertiesJavadocs.get(property);
+            final DetailNode currentModuleJavadoc = propertiesJavadocs.get(currentModuleName);
+            writePropertyRow(sink, property, propertyJavadoc, instance, currentModuleJavadoc);
+        }
+    }
+
+    /**
+     * Writes a table row with 5 columns for the given property - name, description, type,
+     * default value, since.
+     *
+     * @param sink sink to write to.
+     * @param propertyName the name of the property.
+     * @param propertyJavadoc the Javadoc of the property.
+     * @param instance the instance of the module.
+     * @param moduleJavadoc the Javadoc of the module.
+     * @throws MacroExecutionException if an error occurs during writing.
+     */
+    private static void writePropertyRow(Sink sink, String propertyName,
+                                         DetailNode propertyJavadoc, Object instance,
+                                            DetailNode moduleJavadoc)
```

Comment:
too late to do no w ..... just a note for future improvement in design:
there should be no dependency on DetailNode, all this data should be taken before and stored in some POJO objects, and serializer/formatter to xdoc sink just operates on simple objects to put data in final stream.
The same as we have in release notes. One module create structure/object by some logic, Another module do not care how value a calculated and simply place them in result file and care on formatting only.

Reply:
Let's create issue for general improvements, and discuss there. 

> POJO

I was thinking the same, to keep all extraction in one method and then deal with only what we need to avoid `isAssignable` checks everywhere, etc.

Reply:
https://github.com/checkstyle/checkstyle/issues/13673